### PR TITLE
MAINT: optimize.newton: converged=False when secant has zero slope

### LINKED ISSUE #14486: One bug, one mistake and one refactorization proposal for the Secant method implemented in scipy.optimize.zeros.py
My issue is about the implementation of the Secant method in scipy.optimize.zeros.py (within the newton() function).

I am new to GitHub and versioning/pull requests etcetera, so please accept my apologies if I bring this in the wrong way. I am in the process of writing a tutorial on using several scipy submodules (for a course I teach on Scientific Programming) and noticed that `scipy.optimize.root_scalar()` produced different results than the Secant method that I had implemented myself to illustrate the steps. Even worse: `.root_scalar()` reported that the method converged even for the example in which it clearly diverges to -infinity. Hence I took a look at the source code (downloaded from 
https://github.com/scipy/scipy/blob/master/scipy/optimize/zeros.py on Tuesday, July 27, 2021).

1. I found the bug erroneously setting `sol.flag` to `'converged'`: this seems to be a basic typo/copy-past error: `_ECONVERGED` should be replaced by `_ECONVERR` on line 344.
2. I also saw that the code includes some tests on `abs(q0)` versus `abs(q1)`. On lines 346-349, these tests seem to be completely redundant as the formulas used in either case are completely equivalent and should yield the same results (up to truncation errors).
3. On lines 330-331 (which are the cause of the difference between my results and those from `.root_scalar()`), I do see what the programmer was trying to do, but he/she seems to have made a mistake: I'm pretty sure the condition should be `if abs(q1) > abs(q0):` instead of `if abs(q1) < abs(q0):` to make `(p1,q1)` be the "better" (in the sense that `q1` is closer to zero) pair of the two initial guesses. However, if this correction/reordering would really be appropriate (*which I don't think it necessarily is in general; In fact, I can come up with a scenario in which it would be worse*), it would make more sense to implement that correction within the iteration loop (choosing the one(s) from `q0`, `q1` and possibly the newest `f(p)` that are closest to zero). I may be wrong, but I guess that's what the programmer had in mind with lines 346-349, but as I wrote above, those lines seem senseless anyway (*unless it is really just about truncation errors; I'm not really an expert on that, but as far as I can tell, the extra test is not worth it*).

#### Reproducing code example:

A minimal example illustrating the most fundamental bug is the following:
```
def lhs(x):
    return x * np.exp(-x*x) - 0.07

optim.root_scalar(lhs, method='secant', x0 = -0.15, x1 = 1.0 )
```

#### Error message:

Actually, the issue is not so much that I get an error message, but rather that I don't. For the initial conditions as specified, the Secant method is supposed to diverge to -infinity. Indeed, the `sol.root` that is reported at `-288782083.8727057` is in line with this divergence. However, `sol.flag` is erroneously set to `'converged'` and `sol.converged` to `True`.

```
C:\ProgramData\Anaconda3\lib\site-packages\scipy\optimize\zeros.py:343: RuntimeWarning: Tolerance of -577564088.0186934 reached.
  warnings.warn(msg, RuntimeWarning)
Out[33]: 
      converged: True
           flag: 'converged'
 function_calls: 10
     iterations: 9
           root: -288782083.8727057
```

#### Scipy/Numpy/Python version information:

```
1.4.1 1.18.1 sys.version_info(major=3, minor=7, micro=6, releaselevel='final', serial=0)
```

#### Proposed solution:
My proposal is:
1. correct the bug on line 344, replacing `_ECONVERGED` by `_ECONVERR`
2. remove all tests relating to `abs(q0)` versus `abs(q1)` as they just take up time and don't necessarily improve performance, replacing the formula for the next iterate by one that applies in either case.

I have already made these changes and include the relevant section below, so someone could copy-paste it right into the source code, replacing lines 317-356 (*I could also provide a ZIP-file with the code already inserted into zeros.py, and some of my tests, if that would be more convenient*):

```
        # Secant method
        # validate input argument x1
        if x1 is not None:
            if x1 == x0:
                raise ValueError("x1 and x0 must be different")
            p1 = x1
        else: # or use default shift
            eps = 1e-4
            p1 = x0 * (1 + eps)
            p1 += (eps if p1 >= 0 else -eps)
        # evaluate function values q0,q1 at p0,p1
        q0 = func(p0, *args)
        funcalls += 1
        q1 = func(p1, *args)
        funcalls += 1

        for itr in range(maxiter):
            if q1 == q0:
                # Exceptional case: failure to converge due to equal func vals
                # (also catches divergence to infinity or NaN)
                if p1 != p0:
                    msg = "Tolerance of %s reached." % (p1 - p0)
                    if disp:
                        msg += (
                            " Failed to converge after %d iterations, value is %s."
                            % (itr + 1, p1))
                        raise RuntimeError(msg)
                    warnings.warn(msg, RuntimeWarning)
                p = (p1 + p0) / 2.0
                return _results_select(
                    full_output, (p, funcalls, itr + 1, _ECONVERR))
            else:
                # Normal case: next iteration step
                p = p1 + q1 * (p1 - p0) / (q0 - q1)

            # converged to solution within tolerance:
            if np.isclose(p, p1, rtol=rtol, atol=tol):
                return _results_select(
                    full_output, (p, funcalls, itr + 1, _ECONVERGED))

            # prepare for next iteration step, renaming new iterates as current
            p0, q0 = p1, q1
            p1 = p
            q1 = func(p1, *args)
            funcalls += 1
```

P.S. Thinking about it further: I can imagine one reason why the "redundant" check in lines 346-349 might be on purpose after all: because of the effect of truncation errors when dividing by the small number `q0 - q1` (I am a mathematician, not a computer scientist, so truncation errors are not really my field of expertise). However, with the (few) tests) that I did, I did not see any significant differences as a result of changing this line. In the `_array_newton()` function `p = p1 + q1 * (p1 - p0) / (q0 - q1)` is used as well.

In any case, the first point, replacing `_ECONVERGED` by `_ECONVERR` on line 344, still stands. And the reordering of the initial guesses on line 330-331, doesn't necessarily improve things in general at all, by one iteration only at best, so that had better be left to the caller of the method anyway, IMHO.

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-14486

#### What does this implement/fix?
gh-14486 reported that `scipy.optimize.newton`'s secant method reports success even when it fails to converge due to a secant with zero slope. This fixes the issue by implementing the proposed fix: replacing `_ECONVERGED` with `_ECONVERR`.

### CODE REVIEW

File: scipy/optimize/tests/test_zeros.py [hunk 1]
```diff
@@ -809,6 +809,30 @@ def f(x):
...
+        res = root_scalar(lhs, method='secant', x0=-0.15, x1=1.0)
+    assert not res.converged
+    assert res.flag == 'convergence error'
+
+    with pytest.warns(RuntimeWarning, match='Tolerance of'):
+        res = newton(lhs, x0=-0.15, x1=1.0, disp=False, full_output=True)[1]
+    assert not res.converged
+    assert res.flag == 'convergence error'
+
+    # This test case doesn't fail to converge for the vectorized version of
```
Comment (pvanmulbregt) [inline scipy/optimize/tests/test_zeros.py:827]: There are two "this" in the statement. Does one refer to the function `lhs` above perhaps?

File: scipy/optimize/tests/test_zeros.py [hunk 2]
```diff
@@ -809,6 +809,30 @@ def f(x):
...
+    # newton, but this one from `def test_zero_der_nz_dp()` does.
```
Comment (mdhaber) [inline scipy/optimize/tests/test_zeros.py:828]:
```suggestion
    # In the vectorized version of `newton`, the secant method doesn't 
    # encounter zero slope in the problem above. Here's a problem where
    # it does. Check that it reports failure to converge.
```

File: scipy/optimize/tests/test_zeros.py (hunk 1 above)
Comment (mdhaber) [reply scipy/optimize/tests/test_zeros.py:827]: Thanks @pvanmulbregt. Yeah, that was pretty confusing. How does the suggested change look?

File: scipy/optimize/tests/test_zeros.py [hunk 3]
```diff
@@ -809,6 +809,30 @@ def f(x):
...
+    dx = np.finfo(float).eps ** 0.33
+    p0 = (200.0 - dx) / (2.0 + dx)
+    with pytest.warns(RuntimeWarning, match='RMS of'):
+        res = newton(lambda y: (y - 100.0)**2, x0=[p0]*2, full_output=True)
+    assert_equal(res.converged, [False, False])
```
Comment (mdhaber) [inline scipy/optimize/tests/test_zeros.py:833]:
```suggestion
    assert_equal(res.converged, [False, False])
    assert_equal(res.zero_der, [True, True])
```

File: scipy/optimize/tests/test_zeros.py [hunk 4]
```diff
@@ -809,6 +809,32 @@ def f(x):
...
+
+    # In the vectorized version of `newton`, the secant method doesn't 
+    # encounter zero slope in the problem above. Here's a problem where
+    # it does. Check that it reports failure to converge.
+    dx = np.finfo(float).eps ** 0.33
+    p0 = (200.0 - dx) / (2.0 + dx)
+    with pytest.warns(RuntimeWarning, match='RMS of'):
+        res = newton(lambda y: (y - 100.0)**2, x0=[p0]*2, full_output=True)
+    assert_equal(res.converged, [False, False])
+    assert_equal(res.zero_der, [True, True])
```
Comment (mdhaber) [inline scipy/optimize/tests/test_zeros.py:835]:
```suggestion
```

**Review by pvanmulbregt — APPROVED** [review]

**Review by tupui — APPROVED** [review]
LGTM and we have another approval, merging. Thanks Matt and @pvanmulbregt

**Pull request merged**